File: 1388-greatest-sum-divisible-by-three/greatest-sum-divisible-by-three.java
class Solution {
    public int maxSumDivThree(int[] nums) {
        int[][] dp = new int[nums.length][3];
        for(int[] d: dp) {
            Arrays.fill(d, -1);
        }
        return maxSum(nums, 0, 0, dp);
    }
    public int maxSum(int[] nums, int i, int rem, int[][] dp) {
        if(i == nums.length) {
            if(rem == 0) {
                return 0;
            }
            return Integer.MIN_VALUE;
        }
        if(dp[i][rem] != -1) {
            return dp[i][rem];
        }
        int inc = nums[i] + maxSum(nums, i+1, (rem+(nums[i]%3))%3, dp);
        int exc = maxSum(nums, i+1, rem, dp);
        return dp[i][rem] = Math.max(inc, exc);
    }
}

File: 1388-greatest-sum-divisible-by-three/greatest-sum-divisible-by-three.cpp
class Solution {
public:
    int maxSumDivThree(vector<int>& nums) {
        vector<vector<long long>> dp(nums.size(), vector<long long>(3, -1));
        long long ans = maxSum(nums, 0, 0, dp);
        return max(0, (int)ans);
    }

    long long maxSum(vector<int>& nums, int i, int rem, vector<vector<long long>>& dp) {
        if(i == nums.size()) {
            if(rem == 0) {
                return 0;
            }
            return INT_MIN;
        }
        
        if(dp[i][rem] != -1) {
            return dp[i][rem];
        }

        long long pick = nums[i] + maxSum(nums, i+1, (rem+nums[i])%3, dp);
        long long skip = maxSum(nums, i+1, rem, dp);

        return dp[i][rem] = max(pick, skip);
    }
};